### PR TITLE
External interface hessians

### PR DESCRIPTION
**Pull request by carldlaird**
This PR adds support for Hessians to the ExternalGreyBoxModel and the PyomoGreyBoxNLP.

Note: this functionality supports solving a Pyomo model where some of the model constraints come from an external source (e.g., compiled C code). Different from external functions (which support multi-input / single-output), this supports multi-input / multi-output external codes.

## Changes proposed in this PR:
- This PR specifically added Hessian support.
-

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/pynumero/interfaces/pyomo_nlp.py [hunk 1]
```diff
@@ -1018,3 +1117,283 @@ def load_state_into_pyomo(self, bound_multipliers=None):
...
+                    temp = row[mask].copy()
+                    row[mask] = col[mask]
+                    col[mask] = temp
+
+            data_list.append(output_hess.data)
+            irow_list.append(self._output_hess_irow)
+            jcol_list.append(self._output_hess_jcol)
+            # outputs_hess = coo_matrix(
+            #     (outputs_hess.data, (self._outputs_hess_irow, self._outputs_hess_jcol)),
+            #     (self._n_primals, self._n_primals))
```
Comment (jsiirola) [inline pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:1390]: Can this be removed?

File: pyomo/contrib/pynumero/interfaces/pyomo_nlp.py [hunk 2]
```diff
@@ -1018,3 +1117,283 @@ def load_state_into_pyomo(self, bound_multipliers=None):
...
+            # # map the columns from the inputs "u" back to the full primals "x"
+            # eq_hess = coo_matrix(
+            #     (eq_hess.data, (self._eq_hess_irow, self._eq_hess_jcol)),
+            #     (self._n_primals, self._n_primals))
+
+        if self._ex_model.n_outputs() > 0:
+            output_hess = self._ex_model.evaluate_hessian_outputs()
+            if self._output_hess_irow is None:
+                # first time through, let's also check that it is lower triangular
+                if np.any(output_hess.row - output_hess.col < 0):
```
Comment (jsiirola) [inline pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:1368]: Can we just make this:
```
if np.any(output_hess.row < output_hess.col):
```

File: pyomo/contrib/pynumero/interfaces/pyomo_nlp.py [hunk 3]
```diff
@@ -1018,3 +1117,283 @@ def load_state_into_pyomo(self, bound_multipliers=None):
...
+                    row[mask] = col[mask]
+                    col[mask] = temp
+
+            data_list.append(eq_hess.data)
+            irow_list.append(self._eq_hess_irow)
+            jcol_list.append(self._eq_hess_jcol)
+            # # map the columns from the inputs "u" back to the full primals "x"
+            # eq_hess = coo_matrix(
+            #     (eq_hess.data, (self._eq_hess_irow, self._eq_hess_jcol)),
+            #     (self._n_primals, self._n_primals))
```
Comment (jsiirola) [inline pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:1362]: Can this be removed?

File: pyomo/contrib/pynumero/interfaces/pyomo_nlp.py [hunk 4]
```diff
@@ -1018,3 +1117,283 @@ def load_state_into_pyomo(self, bound_multipliers=None):
...
+                # primals ('x')
+                self._eq_hess_irow = row = self._inputs_to_primals_map[eq_hess.row]
+                self._eq_hess_jcol = col = self._inputs_to_primals_map[eq_hess.col]
+
+                # mapping may have made this not lower triangular
+                mask = col > row
+                if np.any(mask):
+                    temp = row[mask].copy()
+                    row[mask] = col[mask]
+                    col[mask] = temp
```
Comment (jsiirola) [inline pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:1354]: The following python idiom avoids a copy (and the 'if' is unnecessary):
```
mask = col > row
row[mask], col[mask] = col[mask], row[mask]
```

File: pyomo/contrib/pynumero/interfaces/pyomo_nlp.py [hunk 5]
```diff
@@ -1018,3 +1117,283 @@ def load_state_into_pyomo(self, bound_multipliers=None):
...
+        irow_list = []
+        jcol_list = []
+
+        # Hw_eq(x)
+        eq_hess = None
+        if self._ex_model.n_equality_constraints() > 0:
+            eq_hess = self._ex_model.evaluate_hessian_equality_constraints()
+            if self._eq_hess_jcol is None:
+                # first time through, let's also check that it is lower triangular
+                if np.any(eq_hess.row - eq_hess.col < 0):
```
Comment (jsiirola) [inline pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:1339]: Can this just be:
```
if np.any(eq_hess.row < eq_hess.col):
```

**Review by jsiirola — COMMENTED** [review]
Lots of questions, but overall I think this looks good.

File: pyomo/contrib/pynumero/interfaces/external_grey_box.py [hunk 6]
```diff
@@ -48,7 +47,8 @@
      (or residuals) that need to be converged, and any outputs that
      are computed from your model. It will also need to provide methods to
      compute the residuals, outputs, and the jacobian of these with respect to
-     the inputs. See the documentation on ExternalGreyBoxModel for more details.
+     the inputs. Implement the methods to evaluate hessians is applicable.
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/external_grey_box.py:50]: fixed.

File: pyomo/contrib/pynumero/interfaces/external_grey_box.py [hunk 7]
```diff
@@ -20,8 +20,7 @@
 
 from ..sparse.block_matrix import BlockMatrix
 
-from six import add_metaclass, itervalues, iteritems
-from six.moves import xrange
+from six import add_metaclass, iteritems
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/external_grey_box.py:23]: fixed.

File: pyomo/contrib/pynumero/interfaces/external_grey_box.py [hunk 8]
```diff
@@ -203,10 +238,38 @@ def evaluate_jacobian_outputs(self):
...
+    #     Compute the product of the output constraint multipliers with the
+    #     hessian of the outputs. E.g., y_o^k is the vector of output
+    #     constraint multipliers from set_output_constraint_multipliers,
+    #     u^k are the vector of inputs from set_inputs, and w_o(u) is the
+    #     function that computes the vector of outputs at the values for
+    #     the input variables. This method must return
+    #     H_o^k = sum_i (y_o^k)_i * grad^2_{uu} w_o(u^k)
+    #     """
+    #     raise NotImplementedError('evaluate_hessian_outputs called '
+    #                               'but not implemented in the derived class.')
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/external_grey_box.py:272]: I moved the documentation of these methods to the class docstring with explanation about when they are used.

File: pyomo/contrib/pynumero/interfaces/pyomo_nlp.py [hunk 9]
```diff
@@ -470,6 +471,19 @@ def __init__(self, pyomo_model):
...
+        """
+        for data in greybox_data:
+            c = data._ex_model.model_capabilities()
+            if (c.n_equality_constraints() > 0 \
+               and not c.supports_jacobian_equality_constraints) \
+               or (c.n_equality_constraints() > 0 \
+               and not c.supports_jacobian_equality_constraints)
+                raise NotImplementedError('PyomoGreyBoxNLP does not support models'
+                                          ' without explicit Jacobian support')
+        """
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:485]: This is a breadcrumb for future development. We should support these at some point.

File: pyomo/contrib/pynumero/interfaces/pyomo_nlp.py (hunk 5 above)
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:1339]: Yep. changed.

File: pyomo/contrib/pynumero/interfaces/pyomo_nlp.py (hunk 4 above)
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:1354]: yep. changed.

File: pyomo/contrib/pynumero/interfaces/pyomo_nlp.py (hunk 3 above)
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:1362]: yep. removed.

File: pyomo/contrib/pynumero/interfaces/pyomo_nlp.py (hunk 2 above)
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:1368]: yep. changed.

File: pyomo/contrib/pynumero/interfaces/pyomo_nlp.py (hunk 1 above)
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/pyomo_nlp.py:1390]: yep. And I changed the np.any stuff above.

File: pyomo/contrib/pynumero/interfaces/tests/compare_utils.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+from pyomo.contrib.pynumero.dependencies import (numpy as np, scipy)
+
+def check_vectors_specific_order(tst, v1, v1order, v2, v2order, v1_v2_map=None):
+    tst.assertEqual(len(v1), len(v1order))
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/tests/compare_utils.py:13]: We should make this kind of comparison general somewhere in pynumero. Can we work on the documentation then?

File: pyomo/contrib/pynumero/interfaces/tests/compare_utils.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    tst.assertEqual(len(v2), len(v2order))
+    tst.assertEqual(len(v1), len(v2))
+    if v1_v2_map is None:
+        v2map = {s:v2order.index(s) for s in v1order}
+    else:
+        v2map = {s:v2order.index(v1_v2_map[s]) for s in v1order}
+    for i,s in enumerate(v1order):
+        tst.assertAlmostEqual(v1[i], v2[v2map[s]], places=7)
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/tests/compare_utils.py:21]: Ok, can we reserve this modification for a generalization of these capabilities somewhere?

File: pyomo/contrib/pynumero/interfaces/tests/compare_utils.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    m2c = np.zeros((len(m2rows), len(m2cols)))
+    if m1_m2_rows_map is None:
+        rowmap = [m2rows.index(x) for x in m1rows]
+    else:
+        rowmap = [m2rows.index(m1_m2_rows_map[x]) for x in m1rows]
+    if m1_m2_cols_map is None:
+        colmap = [m2cols.index(x) for x in m1cols]
+    else:
+        colmap = [m2cols.index(m1_m2_cols_map[x]) for x in m1cols]
+    print(colmap)
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/tests/compare_utils.py:47]: yep. done.

File: pyomo/contrib/pynumero/interfaces/tests/test_dynamic_model.py [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+        irow = np.zeros(nnz, dtype=np.int64)
+        jcol = np.zeros(nnz, dtype=np.int64)
+        data = np.zeros(nnz, dtype=np.float64)
+        idx = 0
+#        # Hess h1bal
+#        for i in range(N-1):
+#            irow[idx] = 2*(N-1)+i+1
+#            jcol[idx] = 2*(N-1)+i+1
+#            data[idx] = lam[i]*dt/A1*(-c1/4)*h1[i+1]**(-1.5)
+#            idx += 1
```
[the remark this answers is not in the document]
Comment (carldlaird) [reply pyomo/contrib/pynumero/interfaces/tests/test_dynamic_model.py:292]: nope. deleted.

**Review by jsiirola — APPROVED** [review]

**Pull request merged**